drivers/hmis/app/models/hmis/hud/project.rb

```diff
@@ -57,7 +57,15 @@ class Hmis::Hud::Project < Hmis::Hud::Base
 
   has_many :services, through: :enrollments_including_wip
   has_many :custom_services, through: :enrollments_including_wip
-  has_many :hmis_services, through: :enrollments_including_wip
+
+  # FIXME: joining services through client_projects confounds postgres. On larger projects, the query might take many
+  # minutes. It needs optimization; for now we pull down the ids
+  # has_many :hmis_services, through: :enrollments_including_wip
+  def hmis_services
+    project = self
+    ids = project.enrollments_including_wip.pluck(:EnrollmentID)
+    Hmis::Hud::HmisService.where(data_source: project.data_source).where(hs_t[:EnrollmentID].in(ids))
+  end
```

Comment:
Yeesh. Did you consider joining with project like this? Not sure if postgres would be equally confused by that.
```
Hmis::Hud::HmisService.joins(:project).where(Hmis::Hud::Project.arel_table[:id].eq(405)).to_sql

=> "SELECT \"hmis_services\".* FROM \"hmis_services\" INNER JOIN \"hmis_client_projects\" ON \"hmis_client_projects\".\"data_source_id\" = \"hmis_services\".\"data_source_id\" AND \"hmis_client_projects\".\"EnrollmentID\" = \"hmis_services\".\"EnrollmentID\" INNER JOIN \"Project\" ON \"Project\".\"DateDeleted\" IS NULL AND \"Project\".\"DateDeleted\" IS NULL AND \"Project\".\"id\" = \"hmis_client_projects\".\"project_id\" WHERE \"hmis_services\".\"DateDeleted\" IS NULL AND \"Project\".\"id\" = 405"
```

for my own comparison, it is different from the main branch has_many query:
```
# main branch has_many
Hmis::Hud::Project.find(405).hmis_services.to_sql

=> "SELECT \"hmis_services\".* FROM \"hmis_services\" INNER JOIN \"Enrollment\" ON \"hmis_services\".\"EnrollmentID\" = \"Enrollment\".\"EnrollmentID\" AND \"hmis_services\".\"PersonalID\" = \"Enrollment\".\"PersonalID\" AND \"hmis_services\".\"data_source_id\" = \"Enrollment\".\"data_source_id\" INNER JOIN \"hmis_client_projects\" ON \"Enrollment\".\"id\" = \"hmis_client_projects\".\"enrollment_id\" WHERE \"hmis_services\".\"DateDeleted\" IS NULL AND \"Enrollment\".\"DateDeleted\" IS NULL AND \"hmis_client_projects\".\"project_id\" = 405"
```

Reply:
> Hmis::Hud::HmisService.joins(:project).where(Hmis::Hud::Project.arel_table[:id].eq(405))

Oh that's much better falling back to pluck and seems to avoid whatever the performance issue is. Thank you!

Reply:
Interestingly this still blows up postgres when we order by DateProvided. I modified the hmis services view to order the results which seems to be enough for postgres to figure it out 🤞

Reply:
Huh. Well, I tested this and it seems to be working, for a combination of Custom and HUD services at WIP and non-WIP enrollments.